src/supernova/mod.rs

```diff
@@ -0,0 +1,1178 @@
+      if U.X.len() != 2 {
+        debug!(
+          "r_U_secondary got instance length {:?} != {:?}",
+          U.X.len(),
+          2
+        );
+        Err(NovaError::ProofVerifyError)
+      } else {
+        Ok(())
+      }
+    })?;
+
+    // check if the output hashes in R1CS instances point to the right running instances
+    let num_field_primary_ro = 3 // params_next, i_new, program_counter_new
+    + 2 * pp.F_arity_primary // zo, z1
+    + (7 + 2 * pp.augmented_circuit_params_primary.get_n_limbs()); // # 1 * (7 + [X0, X1]*#num_limb)
+    let num_field_secondary_ro = 3 // params_next, i_new, program_counter_new
+    + 2 * pp.F_arity_primary // zo, z1
+    + self.num_augmented_circuits * (7 + 2 * pp.augmented_circuit_params_primary.get_n_limbs()); // #num_augmented_circuits * (7 + [X0, X1]*#num_limb)
+
+    let (hash_primary, hash_secondary) = {
+      let mut hasher = <G2 as Group>::RO::new(pp.ro_consts_secondary.clone(), num_field_primary_ro);
+      hasher.absorb(self.pp_digest);
+      hasher.absorb(G1::Scalar::from(self.i as u64));
+      hasher.absorb(self.program_counter);
+
+      for e in z0_primary {
+        hasher.absorb(*e);
+      }
+      for e in &self.zi_primary {
+        hasher.absorb(*e);
+      }
+      if let Some(U) = self.r_U_secondary[0].as_ref() {
+        U.absorb_in_ro(&mut hasher)
+      }
+
+      let mut hasher2 =
+        <G1 as Group>::RO::new(pp.ro_consts_primary.clone(), num_field_secondary_ro);
+      hasher2.absorb(scalar_as_base::<G1>(self.pp_digest));
+      hasher2.absorb(G2::Scalar::from(self.i as u64));
+      hasher2.absorb(G2::Scalar::ZERO);
+      for e in z0_secondary {
+        hasher2.absorb(*e);
+      }
+      for e in &self.zi_secondary {
+        hasher2.absorb(*e);
+      }
+      let default_value = RelaxedR1CSInstance::default(ck_primary, &pp.r1cs_shape_primary);
+      self.r_U_primary.iter().for_each(|U| {
+        U.as_ref()
+          .unwrap_or(&default_value)
+          .absorb_in_ro(&mut hasher2);
+      });
+
+      (
+        hasher.squeeze(NUM_HASH_BITS),
+        hasher2.squeeze(NUM_HASH_BITS),
+      )
+    };
+
+    if hash_primary != self.l_u_secondary.X[0] {
+      debug!(
+        "hash_primary {:?} not equal l_u_secondary.X[0] {:?}",
+        hash_primary, self.l_u_secondary.X[0]
+      );
+      return Err(NovaError::ProofVerifyError);
+    }
+    if hash_secondary != scalar_as_base::<G2>(self.l_u_secondary.X[1]) {
+      debug!(
+        "hash_secondary {:?} not equal l_u_secondary.X[1] {:?}",
+        hash_secondary, self.l_u_secondary.X[1]
+      );
+      return Err(NovaError::ProofVerifyError);
+    }
+
+    // check the satisfiability of the provided `circuit_index` instance
+    let default_instance = RelaxedR1CSInstance::default(ck_primary, &pp.r1cs_shape_primary);
+    let default_witness = RelaxedR1CSWitness::default(&pp.r1cs_shape_primary);
+    let (res_r_primary, (res_r_secondary, res_l_secondary)) = rayon::join(
```

Comment:
In Supernova's paper the verifier need to check every related instance-witness pair (U, W) satisfys  relaxed R1CS constrains. However, it seems only the claimed pair is checked here.

Reply:
They are all verified in their own step. opcode 1 runs x times and is verified, then opcode 2. If the ROM is something like [1,2,1,1] it is run and verified for each step with the code you commented. At the very end a valid SuperNova proof would have both 1 and 2 be valid. 

https://github.com/microsoft/Nova/blob/ea14cabf116b6be34fee112737e00a64d8c81f7c/src/supernova/mod.rs#L1017C1-L1018C102

Reply:
Thanks for your explanation.

Reply:
@wyattbenno777 I'm confused a bit by your explanation. In SuperNova instantiated over a cycle of curves, if there are \ell different step circuits, we will have \ell running instances (on the first curve), a running instance (on the second curve), and a fresh instance (second curve). We need to check that all \ell + 2 instances are satisfying. In the case of Nova, \ell = 1, so we have 3 instances verified in the main branch.

Reply:
In code, we are giving a PrimaryCircuit and a TrivialCircuit to a RunningInstance object. It runs with these as many times as mentioned in the ROM (there is an instance being folded in at each step). In the case of Nova the user is still specifying a primary and a trivial circuit. So should I be explaining this as 3 things to be verified for each opcode? (Primary, Trivial, Thing to be folded on second curve) or as the things the user specifies (Primary and Trivial circuit)?

Reply:
Extra info for any future reader: pg. 15 https://eprint.iacr.org/2022/1758.pdf

Reply:
> They are all verified in their own step. opcode 1 runs x times and is verified, then opcode 2. If the ROM is something like [1,2,1,1] it is run and verified for each step with the code you commented. At the very end a valid SuperNova proof would have both 1 and 2 be valid.
> 
> https://github.com/microsoft/Nova/blob/ea14cabf116b6be34fee112737e00a64d8c81f7c/src/supernova/mod.rs#L1017C1-L1018C102

More addon. I think `opcode 1 runs x times` is indeed a bit confusing haha, because it sounds like we can run opcode 1 x times first then follow by opcode 2 runs y times, ....,  this mindset can explain well on Uniform-IVC Nova, but on Supernova, non-uniform with sequence need to be emphasised. 

Giving example of sequence [1, 2, 1, 1], Supernova assure 
- i=0, augmented circuit 1 is correctly invoked
- i=1, augmented circuit 2 is correctly invoked
- i=2, augmented circuit 1 ...
- i=3, augmented circuit 1 ...

and supernova need to capture soundness from the maliculous prover giving another sequence [1, 1, 1, 2]. Since order matters, and `y1 != y2, for y1=F_1(F_2(F_1(F_1(x)))), y2=F_1(F_1(F_1(F_2(x))))`.


Giving there are N opcodes on 2 cycles curve, there will be N running instances (on second curve) + 1 running instance (on first curve) + 1 fresh instance (second curve)

In each round of folding for the sake of efficiency, we can just verify **One** running instances (on second curve) + 1 running instance (on first curve) + 1 fresh instance (second curve).  The **One** is the opcode being folded in this round. Of course we can verify N running instance (on second curve) all together, but it's not nessesary since they will be verified on their own round. 

Actually, invoking `verify` on RecursiveSNARK (here verify is not the same as NIVC verify circuit!) at each round is more for demo purpose. In production I think we can just update RecursiveSNARK and have a final SNARK prove via CompressSNARK is enough (Terminlogies borrow from nova implementation, not paper)

Reply:
This is a very well thought out answer!

*CycleNova will change this if/when implemented 😊